src/ray/common/asio/instrumented_io_context.cc

```diff
@@ -100,24 +72,63 @@ void instrumented_io_context::post(std::function<void()> handler,
   // the handlers to execute without segfaulting even if the thread that constructed the
   // io_context object exits without joining the handler threads (which can happen in a
   // hard exit).
-  boost::asio::io_context::post([handler = std::move(handler), stats = it->second,
-                                 global_stats = global_stats_, name, start]() {
-    int64_t start_execution = absl::GetCurrentTimeNanos();
-    // Execute actual handler.
-    handler();
-    int64_t end_execution = absl::GetCurrentTimeNanos();
-    // Update queue and execution time stats.
-    const auto queue_time_ns = start_execution - start;
-    const auto execution_time_ns = end_execution - start_execution;
-    // Update handler-specific stats.
-    {
-      absl::MutexLock lock(&(stats->mutex));
-      // Handler-specific execution stats.
-      stats->stats.cum_execution_time += execution_time_ns;
+  boost::asio::io_context::post(
+      [handler = std::move(handler), stats = start_and_stats.second,
+       global_stats = global_stats_, name, start = start_and_stats.first]() {
+        Instrument(std::move(handler), stats, global_stats, start);
+      });
+}
+
+std::pair<int64_t, std::shared_ptr<GuardedHandlerStats>> instrumented_io_context::Enqueue(
+    const std::string name) {
+  auto stats = GetOrCreate(name);
+  auto enqueue_time = Enqueue(stats);
+  return std::make_pair<int64_t, std::shared_ptr<GuardedHandlerStats>>(
+      std::move(enqueue_time), std::move(stats));
+}
+
+int64_t instrumented_io_context::Enqueue(std::shared_ptr<GuardedHandlerStats> stats) {
```

Comment:
Instead of returning an int64, how about returning a unique ptr to an opaque handle that you can pass to Instrument()? The handle can raise an error if not destroyed without being passed there to avoid errors.

Reply:
I thought about doing something like that, but the issue is supporting optimizations such as those used by the periodical runner where it's desirable to `RecordStart()` and `Instrument()` on the same handle many times. I suppose this could work if we also have a `RecordStart()` that takes the opaque handle, and a `GetHandle()` function? So the flow could be:
```c++
auto handle = io_context_.GetHandle("some_handler_name");
io_context_.RecordStart(handle);
io_context_.Instrument(fn, handle);
```
I think with that we could do away with the other `RecordStart()` and `Instrument()` interfaces if we had them all take the opaque handle and if we exposed `GetHandle(handler_name)`, right?